### PR TITLE
Fixes #21962 - switch jquery ajax to axios

### PR DESCRIPTION
**Pull request by amirfefer**
jquery is a enormous library, and there is no point to include it only for ajax calls.
Instead, it is better to use dedicated library for async API calls.
Axios is well known library, and well maintained, with easy mocks implementations for testing, and fits to redux async actions.

- [x] Add axios
- [x] Add mocks implementations for testing
- [x] refactor `ajaxRequestAction` to use axios
- [x]  refactor API  `get` function  to use axios
- [x]  add API `post, put and delete` functions 
- [x]  refactor legacy tests

### CODE REVIEW

File: webpack/assets/javascripts/react_app/API.js [hunk 1]
```diff
@@ -1,6 +1,18 @@
 import $ from 'jquery';
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/API.js]: probably should drop this line (not using jquery?)

File: webpack/assets/javascripts/react_app/API.js [hunk 2]
```diff
@@ -1,6 +1,18 @@
...
+import axios from 'axios';
+
+const onSuccess = response => response.data;
+const onError = error => Promise.reject(error);
 
 export default {
+  ajax(url) {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/API.js]: why expose an "ajax" endpoint? (instead of replacing get, post put etc?)

File: webpack/assets/javascripts/react_app/API.js [hunk 3]
```diff
@@ -1,6 +1,18 @@
 import $ from 'jquery';
+import axios from 'axios';
+
+const onSuccess = response => response.data;
+const onError = error => Promise.reject(error);
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/API.js]: whats the rational for us to recreate all exceptions? e.g. I would expect a library to have predefined good exceptions for regular http failure codes vs we have to handle them ourselves?

File: webpack/assets/javascripts/react_app/API.js (hunk 2 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/API.js]: ajax is a temporarily endpoint, i'll merge it with get soon

File: webpack/assets/javascripts/react_app/API.js [hunk 4]
```diff
@@ -1,37 +1,55 @@
...
+const getcsrfToken = () => {
+  const token = document.querySelector('meta[name="csrf-token"]');
+
+  if (token) {
+    return token.content;
+  }
+  // fail gracefully when no token is found
+  return '';
+};
+const onSuccess = response => response.data;
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/API.js]: Doesn't it mean we are losing all other response fields?
Why consumers can't just use 
```js
get(url).then(({data}) => ... );
```

File: webpack/assets/javascripts/react_app/API.js [hunk 5]
```diff
@@ -1,37 +1,55 @@
...
+const onError = (error) => {
+  /* eslint-disable no-console */
+  console.log(error);
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/API.js]: Do we really want console.log here?

File: webpack/assets/javascripts/react_app/API.js [hunk 6]
```diff
@@ -1,37 +1,55 @@
...
+        'X-Requested-With': 'XMLHttpRequest',
+        'X-CSRF-Token': getcsrfToken(),
+      },
       },
-    });
+    )
+      .then(onSuccess)
+      .catch(onError);
   },
 };
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/API.js]: I think we will also need a `delete` request.

**Review by sharvit — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/API.js (hunk 4 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/API.js]: No it doesn't, data is the whole respond's json body, the only fields we're losing here are headers and axios config, which I don't see any use of them right now.

File: webpack/assets/javascripts/react_app/API.js (hunk 5 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/API.js]: oh snap...  no we don't :) I forgot to delete it

File: webpack/assets/javascripts/react_app/API.js (hunk 6 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/API.js]: sure we do, though right now there's no  use for it

File: webpack/assets/javascripts/react_app/API.js (hunk 4 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/API.js]: Some API's deciding to return data in headers, mostly when it comes to access tokens.

File: webpack/assets/javascripts/react_app/API.js [hunk 7]
```diff
@@ -1,37 +1,40 @@
...
-        /* eslint-disable no-console */
-        console.log(jqXHR);
+  post(url, data = {}, headers = {}) {
+    return axios.post(
+      url, data,
+      {
+        headers: Object.assign(defaultHeaders, headers),
       },
-    });
+    );
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/API.js]: worth adding delete / patch for completion?

File: webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js [hunk 8]
```diff
@@ -83,11 +83,11 @@ export const stateWithUnreadNotifications = immutable({
   },
 });
 
-export const serverResponse = `{ "notifications":[
+export const serverResponse = `{"data": { "notifications":[
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js:86]: why? I don't assume the server response changed?

File: webpack/assets/javascripts/react_app/mockRequests.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+}) => {
+  mockSuccessRequests.forEach((request) => {
+    mock.onGet(request.searchRegex).reply(200, request.response);
+  });
+  mockFailureRequests.forEach((request) => {
+    mock.onGet(request.searchRegex).reply(request.statusCode);
+  });
+};
+
+export const mockPostRequests = ({
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/mockRequests.js]: we should also have a put for notification drawer mark as read updates?

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.test.js [hunk 10]
```diff
@@ -1,8 +1,7 @@
-import nock from 'nock';
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/common/forms.test.js:1]: should probably remove it from package.json too

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 11]
```diff
@@ -25,27 +25,26 @@ const getNotifications = url => (dispatch) => {
 
   if (isDocumentVisible) {
     API.get(url)
-      .done(onGetNotificationsSuccess)
-      .fail(onGetNotificationsFailed)
-      .always(triggerPolling);
-  } else {
-    // document is not visible, keep polling without api call
-    triggerPolling();
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js]: you skipped all of this functionality on purpose ? (skipping if document is not visible and setting up polling?)

File: webpack/assets/javascripts/react_app/redux/actions/statistics/statistics.fixtures.js [hunk 12]
```diff
@@ -1,3 +1,30 @@
+import { mockGetRequests } from '../../../mockRequests';
+
+const mockSuccessRequests = [
+  {
+    path: '/hosts/:id/power',
+    searchRegex: '/hostss/1/power',
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/statistics/statistics.fixtures.js]: typo

**Review by ohadlevy — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js (hunk 8 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/notifications/notifications.fixtures.js:86]: axios returns a new object, the body response encapsulates in data. 
I've extracted it before, as you can see in outdated changes,  though @sharvit  thinks it better to keep the whole axios response object for later use.

**Comment by amirfefer** [comment]
@ohadlevy Thanks for your review
could you have a second look?

**Comment by amirfefer** [comment]
@ohadlevy I've rebased it with the bookmarks changes, could you have another look?

**Comment by amirfefer** [comment]
@danseethaler Would you mind taking a look?

File: webpack/assets/javascripts/react_app/API.js [hunk 13]
```diff
@@ -1,37 +1,51 @@
...
+    return axios.post(
+      url, data,
+      {
+        headers,
       },
-    });
+    );
+  },
+  delete(url, headers = {}) {
+    return axios.post(
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/API.js]: wrong action? (here and patch too)

File: webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.fixtures.js [hunk 14]
```diff
@@ -10,16 +28,20 @@ export const initialState = immutable({
...
 const requestAction = {
   type: types.BOOKMARKS_REQUEST,
   payload: { controller: 'hosts' },
 };
 
 export const onFailureActions = [
-  requestAction,
   {
-    payload: { error: {}, item: { controller: 'hosts' } },
+    type: types.BOOKMARKS_REQUEST,
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.fixtures.js]: Why not use the fixtures like other tests?

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js [hunk 15]
```diff
@@ -49,36 +33,21 @@ const verifyProps = (item, values) => {
...
-          .then(() =>
-            dispatch(addToast({
-              type: 'success',
-              // eslint-disable-next-line no-undef
-              message: Jed.sprintf('%s was successfully created.', __(item)),
-            }))))
-      .catch((error) => {
-        throw error;
-      });
+    API.post(url, values)
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/common/forms.js]: the reason i kept the method in the function header, is that we would like to use the same code path for updating an existing form too.

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.test.js [hunk 16]
```diff
@@ -33,14 +29,10 @@ describe('form actions', () => {
...
   it('on failed response raise exception', () => {
-    nock('http://localhost')
-      .post('/api/resource')
-      .reply(500, { message: 'oh snap' });
-
     const store = mockStore({ resources: [] });
 
     return store
-      .dispatch(submitForm({ values: { a: 1 }, url: 'http://localhost/api/resource', item: 'Resource' }))
+      .dispatch(submitForm({ values: { a: 1 }, url: '/api/500', item: 'Resource' }))
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/common/forms.test.js]: the url was /api/resource and the return was 500, reading this feels like its the wrong url?

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.test.js [hunk 17]
```diff
@@ -49,31 +41,23 @@ describe('form actions', () => {
...
     return store
-      .dispatch(submitForm({ values: { a: 1 }, url: 'http://localhost/api/resource', item: 'Resource' }))
+      .dispatch(submitForm({ values: { a: 1 }, url: '/api/404', item: 'Resource' }))
       .then(() => {
         throw new Error('Should not hit this then block - test was set up incorrectly');
       })
       .catch((error) => {
         expect(error).toBeInstanceOf(SubmissionError);
-        expect(error.errors).toEqual({ _error: ['Error submitting data: 404 Not Found'] });
+        expect(error.errors._error[0]).toMatch(/Error submitting data: 404*/);
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/common/forms.test.js:78]: can we get the actual error code + text?

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 18]
```diff
@@ -13,7 +13,7 @@ import {
 } from '../../../common/sessionStorage';
 import API from '../../../API';
 
-const defaultNotificationsPollingInterval = 10000;
+const defaultNotificationsPollingInterval = 500000;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js]: wrong entry?
for future reference, in your development env just use something like
```
NOTIFICATIONS_POLLING=1000000 foreman start
```

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 19]
```diff
@@ -25,27 +25,29 @@ const getNotifications = url => (dispatch) => {
...
-  function onGetNotificationsSuccess(response) {
+  function onGetNotificationsSuccess({ data }) {
     dispatch({
       type: NOTIFICATIONS_GET_NOTIFICATIONS,
       payload: {
-        notifications: response.notifications,
+        notifications: data.notifications,
       },
     });
+    triggerPolling();
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js]: why did you add it here too? there is no always/done kind of call back ?

File: webpack/assets/javascripts/react_app/redux/actions/statistics/statistics.test.js [hunk 20]
```diff
@@ -2,20 +2,26 @@ import configureMockStore from 'redux-mock-store';
 import thunk from 'redux-thunk';
 import immutable from 'seamless-immutable';
 
-import { requestData, onFailureActions } from './statistics.fixtures';
+import { failedRequestData, successRequestData, onFailureActions, onSuccessActions } from './statistics.fixtures';
 
 import * as actions from './index';
 
 const mockStore = configureMockStore([thunk]);
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/statistics/statistics.test.js:9]: do you need two stores? (mockStore and Store)

**Review by sharvit — APPROVED** [review]
LGTM

**Comment by danseethaler** [comment]
@amirfefer happy to take a look but will be a bit tied up until the 26th 👍

File: webpack/assets/javascripts/react_app/redux/actions/statistics/statistics.test.js (hunk 20 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/statistics/statistics.test.js:9]: `mockStore` is for store's configuration with a middleware, where `store` is for initialization.
(implementation taken from [documentation](http://arnaudbenard.com/redux-mock-store/))

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js (hunk 19 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/notifications/index.js]: there is no a special call back for always / done, though we could use `then` after catch.

File: webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.fixtures.js (hunk 14 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.fixtures.js]: shouldn't we use const types?

Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.fixtures.js]: you should, but if you look at the diff, than it was using it from the requestAction fixture.

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.test.js (hunk 17 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/common/forms.test.js:78]: it looks like  `axios mock adapter`does not return `statusText` property (only status code).

**Comment by amirfefer** [comment]
@ohadlevy thanks again,  would you mind to take another look?

**Comment by mmoll** [comment]
[test foreman]

**Review by danseethaler — APPROVED** [review]
This looks really good @amirfefer! Thanks for your work 👍

File: webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.fixtures.js [hunk 21]
```diff
@@ -10,16 +28,17 @@ export const initialState = immutable({
 });
 
 export const requestData = { url: '/api/bookmarks', controller: 'hosts' };
+export const failRequestData = { url: '/api/bookmarks', controller: 'error' };
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.fixtures.js]: controller should be a real controller? not sure i understand this fixture?

**Comment by amirfefer** [comment]
@ohadlevy I've changed bookmarks fixtures ambiguous error.
I tend to keep the global mock infrastructure

**Comment by waldenraines** [comment]
@ohadlevy any other requests here or can we merge once the tests are green?

**Comment by waldenraines** [comment]
[test katello]

**Comment by waldenraines** [comment]
@amirfefer can you fix codeclimate please?

**Comment by amirfefer** [comment]
@sharvit , @ohadlevy : i've changed the mock requests from global mocking to per test mocking.
could you have another look?

**Review by sharvit — CHANGES_REQUESTED** [review]
@amirfefer when i ran the tests, all of them are passing but there is one test that receive error (and still passing):
https://gist.github.com/sharvit/559dbeec83239a9fc9bd77e39dc2c579

**Comment by amirfefer** [comment]
@sharvit Thanks. I've fixed that test failure

**Review by sharvit — APPROVED** [review]
Thanks @amirfefer! Everything is working and looking great!

**Comment by amirfefer** [comment]
@ohadlevy  Could we merge this?

File: webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js [hunk 22]
```diff
@@ -9,36 +8,42 @@ import {
...
+    const { url, controller, searchRegex } = requestData;
 
-    store.dispatch(actions.getBookmarks(url, controller));
-    expect(store.getActions()).toEqual(onFailureActions);
+    mockRequest({
+      searchRegex,
+      status: 422,
+    });
+    return store.dispatch(actions.getBookmarks(url, controller))
+      .then(() => expect(store.getActions()).toEqual(onFailureActions));
```
Comment (priley86) [inline webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js:32]: maybe there is a better way to avoid promises? I like the approach here overall (and it seems that axios is preferrable to fetch) but I see this bit as a potential limitation to writing more fully scoped tests...will follow up in an email shortly to you guys.

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js [hunk 23]
```diff
@@ -19,25 +19,25 @@ const getNotifications = url => (dispatch) => {
 
   if (isDocumentVisible) {
     API.get(url)
-      .done(onGetNotificationsSuccess)
-      .fail(onGetNotificationsFailed)
-      .always(triggerPolling);
+      .then(onGetNotificationsSuccess)
+      .catch(onGetNotificationsFailed)
```
Comment (priley86) [inline webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:23]: same thing here based on comment above... lot of promises...

File: webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js (hunk 22 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js:32]: @priley86 +1
Checkout this gist:
https://gist.github.com/sharvit/cf658a8494758be8062a63121b925fdc

This is kind of a solution for unit-testing i tried to implement.
We have been discussing about it quit a lot at the last month.
We feel we should migrate the tests soon to this style and start writing integration tests.

Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js:32]: This approach is suggested by redux ( Async Action Creators part):
https://redux.js.org/docs/recipes/WritingTests.html
I agree that we should have some test conventions, especially when it involves integration between react component and redux, I've started to work on that kind of integration helper.

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js (hunk 23 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:23]: What is wrong with promises in API calls? is there a better way?
The second promise here is like `.always` in jquery: 
https://github.com/axios/axios/issues/792

File: webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js (hunk 22 above)
Comment (priley86) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js:32]: Again we have same dilemma....

```
export const submitForm = ({
  item, url, values, method = 'post',
}) => {
  verifyProps(item, values);
  return dispatch =>
    API[method](url, values)
      .then(({ data }) => {
        dispatch({
          type: `${item.toUpperCase()}_FORM_SUBMITTED`,
          payload: { item, data },
        });
```
... I am chaining promises/chaining actions.... do you not prefer it to be `gen.next()` ?

Comment (priley86) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js:32]: I <3 the use of Jest snapshots though.. maybe we can incorporate it with Saga too... just a thought.

File: webpack/assets/javascripts/react_app/redux/actions/notifications/index.js (hunk 23 above)
Comment (priley86) [reply webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:23]: I think the syntax here is much cleaner, personally, but will leave it to you to decide it ;)

https://redux-saga.js.org/docs/advanced/Testing.html

File: webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js (hunk 22 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js:32]: +1 on jest snapshots,
I think we should have a deeper look on saga, particularly if its benefit worth the migration from thunk.  I don't sure if we have complicated API requests (which require chaining promises or so) in foreman, testing would be easier though.

Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js:32]: leaving saga aside, any thoughts on redux-promise middlewere?
It could answer your concern with promises and yet simpler to migrate.

Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js:32]: where does this leave this PR? I'm not sure i fully understand its status?

**Review by ohadlevy — APPROVED** [review]
I believe this is a good step forward, we still have an unresolved discussion around how to test async actions, but imho its outside the scope of this PR. thanks @amirfefer

File: webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js (hunk 22 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/redux/actions/bookmarks/bookmarks.test.js:32]: I don't think it relevant to this PR, @priley86  has mentioned good points that we should look into.
